src/modelmanager.cpp

```diff
@@ -485,47 +502,45 @@ Status ModelManager::reloadModelWithVersions(ModelConfig& config) {
 
     auto model = getModelIfExistCreateElse(config.getName());
     getVersionsToChange(config, model->getModelVersions(), requestedVersions, versionsToStart, versionsToReload, versionsToRetire);
-    downloadModels(fs, config, versionsToStart);
-    downloadModels(fs, config, versionsToReload);
 
-    status = model->addVersions(versionsToStart, config);
-    if (!status.ok()) {
-        spdlog::error("Error occurred while loading model: {} versions; error: {}",
-            config.getName(),
-            status.string());
-        return status;
+    try {
+        downloadModels(fs, config, versionsToStart);
+
+        status = model->addVersions(versionsToStart, config);
+        if (!status.ok()) {
+            spdlog::error("Error occurred while loading model: {} versions; error: {}",
+                config.getName(),
+                status.string());
+        }
+
+        cleanupModelTmpFiles(config, versionsToStart);
+    } catch (std::exception& e) {
+        cleanupModelTmpFiles(config, versionsToStart);
     }
-    status = model->reloadVersions(versionsToReload, config);
-    if (!status.ok()) {
-        spdlog::error("Error occurred while reloading model: {}; versions; error: {}",
-            config.getName(),
-            status.string());
-        return status;
+
+    try {
+        downloadModels(fs, config, versionsToReload);
+        status = model->reloadVersions(versionsToReload, config);
+        if (!status.ok()) {
+            spdlog::error("Error occurred while reloading model: {}; versions; error: {}",
+                config.getName(),
+                status.string());
+        }
+
+        cleanupModelTmpFiles(config, versionsToReload);
```

Comment:
push cleanupModelTmpFiles after the catch. Otherwise you could throw from inside cleanupModelTmpFiles and then call it again.
Please log error inside catch clause.

The same with changes above.